File: app/src/library.ts
// Library / acquisition bridge over Tauri IPC (M1). Mirrors
// crates/dj-library types; falls back to nulls outside Tauri so the UI
// stays testable headless (tests inject a mock client).

import { IpcClient } from './ipc';

export interface LicenseInfo {
  kind: string;
  name: string;
  url: string;
  attribution: string;
}

export interface Track {
  id: number;
  title: string;
  artist: string;
  album: string;
  file_path: string;
  content_hash: string;
  format: string;
  duration_secs: number | null;
  sample_rate: number | null;
  channels: number | null;
  source: string;
  source_ref: string;
  license: LicenseInfo;
  analysis_status: string;
  bpm: number | null;
  musical_key: string | null;
  created_at: string;
  updated_at: string;
}

/** What a delete took with it: the row that is gone, and whether its
 *  audio file went too (only files the app downloaded or rendered do). */
export interface DeletedTrack {
  track: Track;
  file_removed: boolean;
}

export interface TrackResult {
  provider: string;
  /** How this result is acquired. Explicit because a Download provider's
   *  URL may only be resolved at acquire time (e.g. Internet Archive). */
  acquire_kind: 'download' | 'deep_link';
  id: string;
  title: string;
  artist: string;
  album: string;
  duration_secs: number | null;
  preview_url: string | null;
  artwork_url: string | null;
  license: LicenseInfo;
  download_url: string | null;
  deep_link_url: string | null;
}

export interface FilterOption {
  value: string;
  label: string;
}

/** A select-style filter a provider supports (first option = default). */
export interface FilterSpec {
  id: string;
  label: string;
  options: FilterOption[];
}

export interface ProviderInfo {
  id: string;
  name: string;
  acquire_kind: 'download' | 'deep_link';
  filters: FilterSpec[];
}

/** A provider download running in the background (or recently finished). */
export interface DownloadJob {
  id: number;
  provider: string;
  /** Provider-side result id — matches a job to its result row. */
  result_id: string;
  title: string;
  state: 'running' | 'done' | 'failed';
  /** Completed fraction (0..1) when the transfer size is known. */
  fraction: number | null;
  stage: string;
  error: string | null;
  track_id: number | null;
}

/** Analysis queue snapshot (M3): background worker progress. */
export interface AnalysisQueue {
  /** Track id currently being analyzed, if any. */
  current: number | null;
  /** Track ids still waiting, in queue order. */
  queued: number[];
  /** Track counts by analysis status (queued/analyzing/done/failed). */
  counts: Record<string, number>;
}

/** What LibraryView needs; the Tauri-backed client below implements it and
 *  tests substitute a mock. */
export interface LibraryClientApi {
  tracks(): Promise<Track[] | null>;
  search(text: string): Promise<Track[] | null>;
  /** Enabled providers with their per-store filter specs. */
  providers(): Promise<ProviderInfo[] | null>;
  /** Search one store with that store's filter selections. */
  searchProvider(
    provider: string,
    text: string,
    filters: Record<string, string>,
  ): Promise<TrackResult[] | null>;
  importTrack(path: string): Promise<Track | null>;
  /** Rename a track. Returns the row as it now stands (the backend trims
   *  both fields and refuses a blank title). */
  setTrackNames(trackId: number, title: string, artist: string): Promise<Track | null>;
  /** Delete a track: its row and DJ metadata, its cached stems, and the
   *  audio file when the app owns it (a download or a rendered clip). */
  deleteTrack(trackId: number): Promise<DeletedTrack | null>;
  /** Import a rekordbox XML export (M4): tracks/beatgrids/cues/loops. */
  importRekordbox(path: string): Promise<{ imported: number; duplicates: number } | null>;
  /** Queue a download; the transfer runs on a backend thread. Returns the
   *  job id whose progress `downloadJobs` reports. */
  startDownload(result: TrackResult): Promise<number | null>;
  /** Running/recent download jobs (polled while a download is in flight). */
  downloadJobs(): Promise<DownloadJob[] | null>;
  openStorePage(result: TrackResult): Promise<string | null>;
  /** Open a web URL in the system's default browser (never the webview). */
  openExternal(url: string): Promise<void | null>;
  playbackLoad(instance: string, trackId: number): Promise<void | null>;
  /** Background analysis queue snapshot (M3). */
  analysisStatus(): Promise<AnalysisQueue | null>;
  /** Queue (or re-run) analysis for a track. */
  analyzeTrack(trackId: number): Promise<void | null>;
}

export class LibraryClient extends IpcClient implements LibraryClientApi {
  tracks() {
    return this.call<Track[]>('library_tracks');
  }
  search(text: string) {
    return this.call<Track[]>('library_search', { text });
  }
  providers() {
    return this.call<ProviderInfo[]>('providers');
  }
  searchProvider(provider: string, text: string, filters: Record<string, string>) {
    return this.call<TrackResult[]>('search_provider', { provider, text, filters });
  }
  importTrack(path: string) {
    return this.call<Track>('import_track', { path });
  }
  setTrackNames(trackId: number, title: string, artist: string) {
    return this.call<Track>('set_track_names', { trackId, title, artist });
  }
  deleteTrack(trackId: number) {
    return this.call<DeletedTrack>('delete_track', { trackId });
  }
  importRekordbox(path: string) {
    return this.call<{ imported: number; duplicates: number }>('import_rekordbox', { path });
  }
  startDownload(result: TrackResult) {
    return this.call<number>('start_download', { result });
  }
  downloadJobs() {
    return this.call<DownloadJob[]>('download_jobs');
  }
  openStorePage(result: TrackResult) {
    return this.call<string>('open_store_page', { result });
  }
  async openExternal(url: string) {
    await this.ready;
    if (!this.invoke) {
      // Plain-browser dev mode: a new tab is the "default browser" here.
      window.open(url, '_blank', 'noopener,noreferrer');
      return null;
    }
    return this.call<void>('open_external', { url });
  }
  playbackLoad(instance: string, trackId: number) {
    return this.call<void>('playback_load', { instance, trackId });
  }
  analysisStatus() {
    return this.call<AnalysisQueue>('analysis_status');
  }
  analyzeTrack(trackId: number) {
    return this.call<void>('analyze_track', { trackId });
  }
}

export const library = new LibraryClient();

File: app/src-tauri/src/library.rs
//! Library + acquisition IPC (M1): the track library, provider search
//! and downloads, watch folders, and the Playback/Audio modules' track
//! loading — everything the Library page and the track-playing panels
//! call.

use dj_engine::audio::AudioStatus;
use dj_library::{ProviderInfo, Query, Track, TrackResult};
use serde::Serialize;
use std::collections::BTreeMap;
use std::path::{Path, PathBuf};
use tauri::State;

use crate::{engine_lock, err, patch_edit, AppState, CmdError, CmdResult, EditKey};

#[tauri::command]
pub(crate) fn library_tracks(state: State<AppState>) -> CmdResult<Vec<Track>> {
    state.library.tracks().map_err(err)
}

#[tauri::command]
pub(crate) fn library_search(state: State<AppState>, text: String) -> CmdResult<Vec<Track>> {
    state.library.search(&text).map_err(err)
}

/// Enabled acquisition providers with their UI filter specs (per-store
/// search tabs).
#[tauri::command]
pub(crate) fn providers(state: State<AppState>) -> CmdResult<Vec<ProviderInfo>> {
    Ok(state.hub.providers_info())
}

/// Search one store, with that store's filter selections. Runs on a
/// worker thread (`async`): a provider search is a network round-trip, and
/// the YouTube one spawns yt-dlp — neither may block the main thread.
#[tauri::command(async)]
pub(crate) fn search_provider(
    state: State<AppState>,
    provider: String,
    text: String,
    filters: BTreeMap<String, String>,
) -> CmdResult<Vec<TrackResult>> {
    let mut q = Query::new(&text);
    q.filters = filters;
    state.hub.search_provider(&provider, &q).map_err(err)
}

#[tauri::command]
pub(crate) fn import_track(state: State<AppState>, path: String) -> CmdResult<Track> {
    state
        .library
        .import_file(&PathBuf::from(path), dj_library::ImportOptions::default())
        .map(|o| o.track().clone())
        .map_err(err)
}

/// Rename a track (Library page): title and artist only — everything
/// else about the row is derived from the file or the analysis.
#[tauri::command]
pub(crate) fn set_track_names(
    state: State<AppState>,
    track_id: i64,
    title: String,
    artist: String,
) -> CmdResult<Track> {
    state
        .library
        .set_track_names(track_id, &title, &artist)
        .map_err(err)
}

/// Delete a track and everything the app derived from it: the row with
/// its DJ metadata (cues, loops, beatgrid, tags, crate membership), the
/// stem cache keyed by its content, the Clip page's decoded copy, and the
/// audio file itself when the app owns it (a provider download or a
/// rendered clip — a file in the user's own folders always stays).
///
/// It does NOT chase the references pointing AT the track: a beat clip's
/// source and a saved patch's deck each name a track that may already be
/// gone, and each degrades on its own (a clip keeps its audio and says
/// its source is unknown, a patch load warns and comes up without the
/// track).
#[tauri::command]
pub(crate) fn delete_track(state: State<AppState>, track_id: i64) -> CmdResult<dj_library::DeletedTrack> {
    // A separation in flight would write stems back into the cache we are
    // about to remove, so stop it before the row goes.
    state.stems.cancel_track(track_id);
    let deleted = state.library.delete_track(track_id).map_err(err)?;
    state.clips.forget(track_id);
    let hash = &deleted.track.content_hash;
    if let Err(e) = dj_analysis::remove_stems(state.library.data_dir(), hash) {
        eprintln!("[dj-analysis] removing stems for {hash}: {e:#}");
    }
    Ok(deleted)
}

#[derive(Serialize)]
pub(crate) struct RekordboxImportSummary {
    imported: usize,
    duplicates: usize,
}

/// Import a rekordbox XML export (M4, PRD §8.1): tracks, beatgrids, hot
/// cues, and loops land in the library DB; existing tracks (by path) skip.
#[tauri::command]
pub(crate) fn import_rekordbox(state: State<AppState>, path: String) -> CmdResult<RekordboxImportSummary> {
    let report = state
        .library
        .import_rekordbox_xml(Path::new(&path))
        .map_err(err)?;
    Ok(RekordboxImportSummary {
        imported: report.imported.len(),
        duplicates: report.duplicates.len(),
    })
}

/// Start acquiring a result in the background and return the job id. The
/// transfer never runs on this thread: yt-dlp fetches take seconds to
/// minutes, and even HTTP downloads would stall the window.
#[tauri::command]
pub(crate) fn start_download(state: State<AppState>, result: TrackResult) -> CmdResult<u64> {
    Ok(state.downloads.start(result))
}

/// Snapshot of running/recent download jobs (polled by the library view).
#[tauri::command]
pub(crate) fn download_jobs(state: State<AppState>) -> CmdResult<Vec<dj_library::DownloadJob>> {
    Ok(state.downloads.jobs())
}

/// Deep-link acquisition: resolves the store URL, opens it in the system
/// browser, and returns it (M1: iTunes purchases land via the watch folder).
#[tauri::command]
pub(crate) fn open_store_page(state: State<AppState>, result: TrackResult) -> CmdResult<String> {
    state
        .hub
        .open_deep_link(&result, |url| {
            open::that_detached(url).map_err(anyhow::Error::from)
        })
        .map_err(err)
}

/// Open a web URL in the system's default browser (never in the app's
/// webview). Restricted to http(s) so IPC can't be used to launch
/// arbitrary local files/schemes.
#[tauri::command]
pub(crate) fn open_external(url: String) -> CmdResult<()> {
    if !url.starts_with("https://") && !url.starts_with("http://") {
        return Err(CmdError::invalid(format!(
            "refusing to open non-http(s) URL: {url}"
        )));
    }
    open::that_detached(&url).map_err(err)
}

#[tauri::command]
pub(crate) fn add_watch_folder(state: State<AppState>, path: String) -> CmdResult<()> {
    state
        .library
        .add_watch_folder(&PathBuf::from(path))
        .map_err(err)
}

#[tauri::command]
pub(crate) fn watch_folders(state: State<AppState>) -> CmdResult<Vec<String>> {
    Ok(state
        .library
        .watch_folders()
        .map_err(err)?
        .into_iter()
        .map(|p| p.to_string_lossy().to_string())
        .collect())
}

/// Load a library track into a Playback module instance.
#[tauri::command]
pub(crate) fn playback_load(state: State<AppState>, instance: String, track_id: i64) -> CmdResult<()> {
    let track = state.library.track(track_id).map_err(err)?;
    let mut engine = patch_edit(&state, EditKey::Track(&instance))?;
    engine
        .playback_load(&instance, &PathBuf::from(track.file_path))
        .map_err(err)
}

/// Load a library track into an Audio module instance. The track's tempo
/// comes from the library (canonical, like deck beatgrids): the module
/// adopts it on the BPM input and resets speed to 1x.
#[tauri::command]
pub(crate) fn audio_load(state: State<AppState>, instance: String, track_id: i64) -> CmdResult<()> {
    let track = state.library.track(track_id).map_err(err)?;
    let mut engine = patch_edit(&state, EditKey::Track(&instance))?;
    engine
        .audio_load(&instance, &PathBuf::from(track.file_path), track.bpm)
        .map_err(err)
}

/// Track + transport + tempo snapshot for the Audio module panel.
#[tauri::command]
pub(crate) fn audio_status(
    state: State<AppState>,
    instance: String,
) -> CmdResult<AudioStatus> {
    let engine = engine_lock(&state)?;
    engine.audio_status(&instance).map_err(err)
}

/// Waveform overview peaks (0..=1) of an Audio module's track.
#[tauri::command]
pub(crate) fn audio_waveform(state: State<AppState>, instance: String, buckets: usize) -> CmdResult<Vec<f32>> {
    let engine = engine_lock(&state)?;
    engine
        .audio_waveform(&instance, buckets.min(20_000))
        .map_err(err)
}
